### ISSUE TITLE
FEATURE REQUEST:  Saved GQL with client REST to GraphQL benefits!

### ISSUE THREAD

**Issue by corepay** [hasura/graphql-engine#3488]
### Concept:

* I create a query, mutation or subscription (QMS) using the console.  
* When satisfied I have the option to SAVE the QMS.  
* When saved pressed a popup allows me to provide a name for the QMS.
* When name saved, the console stores the QSM (like it does now in HISTORY tab) and assigns a unique HASH.
* The CONSOLE has a new tab showing the saved QSM that can be selected and displayed in console like History does now.

Now that Hasura is saving the QSM for me I should be able to simply send the HASH in cURL or using a REST client or via REST from my app.

Benefits:

1. Allows me save QSMs and easily retrieve them from the console instead of scrolling through hundreds of QSMs created when developing the QSM in the current HISTORY tab
2.  I can use a saved QSM easily in my app without setting up a graphQL client and writing (duplicating) a lot of the QSM is my client app
3.  Using Hasura, I really do not need to learn how to develop grapqhl or write complicated resolvers or schemas on SERVER or IN MY APP - I just build my data structure in the console and use Hasura to create my QSM in the Explorer, grab the hash for the QSM and use already familiar REST methods to integrate with Hasura
4.  BIG game changer -> Since my app is using a hash to a QSM stored on Hasura I can change the structure/parameters on the QSM in Hasura and DO NOT need to change the code in my app.  

The hash could be sent as a header value in all cases or as a path param on GET, UPDATE or a query param on POST.

Ideally, this feature would be part of an SDK to make it even more simple to integrate Hasura on the client side.  

Let me know what you guys think.  Not sure I can help with any PR but feel free to reach out if you want some more thought, opinion or assistance on the use or business case.

**Comment by coco98** [comment]
@corepay Love it! Thanks for writing this up :) 

I think it might be nicer to allow the creation of a named REST endpoint instead of a hash? I like this more because the REST endpoint can have a descriptive user-friend name.

**Comment by corepay** [comment]
IMO you can probably easily have an option to create a descriptive REST endpoint for single-focused context like Users or Items 

But a context named REST endpoint like `/users` wont apply in many cases where I build a QSM across multiple contexts - for example, your 3-factor app.  In this case it would be similar to an RPC call….but typing this the thought occurred that I suppose you can make the endpoint reflect the service rather than the entity.

Sitting here thinking if I had this option I would probably do a mix in my apps.  I would have a single endpoint for my 3-factor-app type requests and entity named endpoints where I am not doing a bunch of stuff outside normal CRUD.

Like to keep in the loop on this if you move forward.  FYI my idea where this would have a big impact is in developer portals with multiple devs and firebase-like Baas....

**Comment by corepay** [comment]
...also just thought about Events.  You can provide pre-built QSM in Events and probably even allow chained events where and event calls the next hashed QSM and executes the query or mutation.

whaddaythinkaboutthat?

**Comment by corepay** [comment]
:) also just thought - you can extend permissions to the QSM in Hasura in total - for example if the QSM crosses users, posts, authors you can set permissions for the hash or named endpoint rather than for each entity (user/post/author)

**Comment by thefliik** [comment]
Usually I've seen this feature referred to as *persisted queries* in the GraphQL world. For some prior art, see https://graphql-ruby.org/operation_store/overview.html#what-are-persisted-queries.

I'll note that persisted queries are usually saved by the client (rather than an admin creating a "special endpoint" in the Hasura console). Letting clients save their own persisted queries would be a better abstraction than what is described here. Though I could imagine that the Hasura console might provide a nice GUI allowing an admin to save a query.

I'll note that a major reason for using persisted queries, not listed above, is reducing the payload needed to trigger a vary large query (using a persisted query, you just need to provide a URL rather than an entire mutation).

**Comment by corepay** [comment]
Good point on the payload and I will check out persisted queries.  One point I thought about maybe not stressed or explained well enough the value of having the query on server side is that it can be changed (within some limits) on the fly without having to change the client side code - as always a big plus for mobile apps and delays in app stores.

**Comment by thefliik** [comment]
I'll also note that Apollo has support for persisted queries: https://blog.apollographql.com/persisted-graphql-queries-with-apollo-client-119fd7e6bba5

**Comment by guyf** [comment]
👍

**Comment by Bessonov** [comment]
Don't miss the linked feature request above. It could be a very good combination.
